src/vm/analysis/update_expressions_sorting/mod.rs

```diff
@@ -0,0 +1,216 @@
+                let exp = &contract_analysis.expressions[*i];
+                self.find_expression_definition(&exp).unwrap()
+            }).collect();
+            let functions_names = deps_props.iter().map(|i| i.0.clone()).collect();
+            let exprs = deps_props.iter().map(|i| i.2.clone()).collect();
+
+            let mut error = CheckError::new(CheckErrors::CyclingDependencies(functions_names));
+            error.set_expressions(exprs);
+            return Err(error)
+        }
+
+        contract_analysis.top_level_expression_sorting = Some(sorted_indexes);
+        Ok(())
+    }
+
+    fn register_dependencies(&mut self, expr: &SymbolicExpression, tle_index: usize) -> CheckResult<bool> {
+        match expr.expr {
+            AtomValue(_) => Ok(true),
+            Atom(ref name) => {
+                if let Some(dep) = self.top_level_expressions_map.get(&name.clone()) {
+                    if dep.atom_index != expr.id {
+                        self.graph.add_directed_edge(tle_index, dep.expr_index)?;
+                    }
+                }
+                Ok(true)
+            },
+            List(ref exprs) => {
+                for expr in exprs.into_iter() {
+                    self.register_dependencies(expr, tle_index)?;
+                }
+                Ok(true)
+            }
+        }
+    }
+
+    fn find_expression_definition<'b>(&mut self, exp: &'b SymbolicExpression) -> Option<(String, u64, &'b SymbolicExpression)> {
+        if let Some(expression) = exp.match_list() {
+            if let Some((function_name, function_args)) = expression.split_first() {
+                if let Some(definition_type) = function_name.match_atom() {
+                    match definition_type.as_str() {
+                        "define-map" | "define-data-var" | "define" | "define-public" | "define-read-only" => {
+                            if function_args.len() > 1 {
+                                let define_expr = match function_args[0].match_list() {
+                                    Some(list) => &list[0],
+                                    _ => &function_args[0]
+                                };
+                                if let Some(tle_name) = define_expr.match_atom() {
+                                    return Some((tle_name.clone(), define_expr.id, define_expr));
+                                }   
+                            }
+                        }
+                        _ => {}
+                    }
+                } 
+            } 
+        }
+        None
+    }
+}
+
+pub struct TopLevelExpressionIndex {
+    expr_index: usize,
+    atom_index: u64
+}
+
+struct Graph {
+    nodes: Vec<usize>,
+    adjacency_list: Vec<Vec<usize>>
+}
+
+impl Graph {
+    fn new() -> Self {
+        Self { 
+            nodes: Vec::new(), 
+            adjacency_list: Vec::new() 
+        }
+    }
+
+    fn add_node(&mut self, expr_index: usize) -> CheckResult<()> {
```

Comment:
Since `add_node` and `add_directed_edge` don't have any error cases (they would panic if something went wrong), they should probably just be void functions.

Reply:
Indeed! Initially, the graph was based on expressions, and at some point, I simplified it without re-evaluating this struct.
Addressed in https://github.com/blockstack/blockstack-core/pull/1079/commits/7ff87ee540f8644628d02b3deb2eaa86d682994a